sklearn/cluster/optics_.py

```diff
@@ -0,0 +1,770 @@
+        Distance at which each sample becomes a core point.
+        Points which will never be core have a distance of inf.
+
+    References
+    ----------
+    Ankerst, Mihael, Markus M. Breunig, Hans-Peter Kriegel, and Jörg Sander.
+    "OPTICS: ordering points to identify the clustering structure." ACM SIGMOD
+    Record 28, no. 2 (1999): 49-60.
+    """
+
+    def __init__(self, min_samples=5, max_bound=np.inf, metric='euclidean',
+                 algorithm='ball_tree', leaf_size=30, p=2,
+                 metric_params=None, maxima_ratio=.75,
+                 rejection_ratio=.7, similarity_threshold=0.4,
+                 significant_min=.003, min_cluster_size_ratio=.005,
+                 min_maxima_ratio=0.001, n_jobs=1):
+
+        self._init_params(n_neighbors=min_samples,
+                          algorithm=algorithm,
+                          leaf_size=leaf_size, metric=metric, p=p,
+                          metric_params=metric_params, n_jobs=n_jobs)
+
+        self.max_bound = max_bound
+        # min_samples is set via n_neighbors attribute in init params
+        self.maxima_ratio = maxima_ratio
+        self.rejection_ratio = rejection_ratio
+        self.similarity_threshold = similarity_threshold
+        self.significant_min = significant_min
+        self.min_cluster_size_ratio = min_cluster_size_ratio
+        self.min_maxima_ratio = min_maxima_ratio
+
+    @property
+    def n_neighbors(self):
+        return self.min_samples
+
+    @n_neighbors.setter
+    def n_neighbors(self, value):
+        self.min_samples = value
+
+    def fit(self, X, y=None):
+        """Perform OPTICS clustering
+
+        Extracts an ordered list of points and reachability distances, and
+        performs initial clustering using `max_bound` distance specified at
+        OPTICS object instantiation.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features)
+            The data.
+
+        Returns
+        -------
+        self : instance of OPTICS
+            The instance.
+        """
+        X = check_array(X, dtype=np.float)
+
+        n_samples = len(X)
+        # Start all points as 'unprocessed' ##
+        self._processed = np.zeros((n_samples, 1), dtype=bool)
+        self.reachability_ = np.ones(n_samples) * np.inf
+        self.core_dists_ = np.ones(n_samples) * np.nan
+        # Start all points as noise ##
+        self.labels_ = -np.ones(n_samples, dtype=int)
+        self.ordering_ = []
+
+        # Check for valid n_samples relative to min_samples
+        if self.min_samples > n_samples:
+            raise ValueError("Number of training samples (n_samples=%d) must "
+                             "be greater than min_samples (min_samples=%d) "
+                             "used for clustering." %
+                             (n_samples, self.min_samples))
+
+        super(OPTICS, self).fit(X)
+
+        self.core_dists_[:] = self.kneighbors(X, self.min_samples)[0][:, -1]
```

Comment:
They do. The criteria for a cluster is min_samples within epsilon distance; the core_dist is the distance at which there are min_samples at that distance, so you want the query point included in that count. (since X is being intersected with the fitted tree of kneighbors, we're asking for neighbors from points that just happen to fall exactly on the nearest neighbor-- so the closest neighbor always has a distance of zero, but is still included).